tests/build/test_utils.py

```diff
@@ -108,3 +111,110 @@ def test_process_accelerate_launch_custom_config_file(patch_path_exists):
     temp_job_config = {"accelerate_launch_args": {"config_file": dummy_config_path}}
     args = process_accelerate_launch_args(temp_job_config)
     assert args.config_file == dummy_config_path
+
+
+class CopyCheckpointTestConfig:
+
+    def __init__(self, test_root):
+
+        assert os.path.isdir(test_root)
+
+        self.test_root = test_root
+        self.source_dir = os.path.join(self.test_root, "test_copytree_source")
+        self.source_sub_dir = os.path.join(self.source_dir, "subdir1")
+
+        if os.path.isdir(self.source_dir):
+            shutil.rmtree(self.source_dir)
+
+        os.mkdir(self.source_dir)
+        open(os.path.join(self.source_dir, "tf1.txt"), "a").close()
+        open(os.path.join(self.source_dir, "tf2.txt"), "a").close()
+        open(os.path.join(self.source_dir, "tf3.txt"), "a").close()
+
+        os.mkdir(self.source_sub_dir)
+        open(os.path.join(self.source_sub_dir, "tf4.txt"), "a").close()
+        open(os.path.join(self.source_sub_dir, "tf5.txt"), "a").close()
+        open(os.path.join(self.source_sub_dir, "tf6.txt"), "a").close()
+
+    def are_dir_trees_equal(self, dir1, dir2):
+
+        dirs_cmp = filecmp.dircmp(dir1, dir2)
+        if (
+            len(dirs_cmp.left_only) > 0
+            or len(dirs_cmp.right_only) > 0
+            or len(dirs_cmp.funny_files) > 0
```

Comment:
Can we also check that `len(dirs_cmp.common_files) > 0`, when I test out `filecmp.dircmp(dir1, dir2)` it seems to only list the first level of files and not subdirs

Reply:
Hmm.... `dirs_cmp.left_only` and `dirs_cmp.right_only` should include both files and subdirectories, and this gets called recursively throughout the file tree. 

In the event any level (including the topmost) are not equal, this bombs out at that level with a False:

Output from a two level check where everything is equal:
```
validating /tmp/test_copytree_source against /tmp/test_copytree_target
	dirs_cmp.right_only: []
	dirs_cmp.left_only: []
validating /tmp/test_copytree_source/subdir1 against /tmp/test_copytree_target/subdir1
	dirs_cmp.right_only: []
	dirs_cmp.left_only: []
True
```

Output from a two level check where the top level fails:
```
validating /tmp/test_copytree_source against /tmp/test_copytree_target
	dirs_cmp.right_only: []
	dirs_cmp.left_only: ['subdir1', 'tf1.txt', 'tf2.txt', 'tf3.txt']
False
```

Code with the debug prints:
```
def are_dir_trees_equal(dir1, dir2):
    print(f'validating {dir1} against {dir2}').       #OUTPUT
    dirs_cmp = filecmp.dircmp(dir1, dir2)
    print(f'\tdirs_cmp.right_only: {dirs_cmp.right_only}').  #OUTPUT
    print(f'\tdirs_cmp.left_only: {dirs_cmp.left_only}').      #OUTPUT

    if (
        len(dirs_cmp.left_only) > 0
        or len(dirs_cmp.right_only) > 0
        or len(dirs_cmp.funny_files) > 0
    ):
        return False
    (_, mismatch, errors) = filecmp.cmpfiles(
        dir1, dir2, dirs_cmp.common_files, shallow=False
    )
    if len(mismatch) > 0 or len(errors) > 0:
        return False
    for common_dir in dirs_cmp.common_dirs:
        new_dir1 = os.path.join(dir1, common_dir)
        new_dir2 = os.path.join(dir2, common_dir)
        if not are_dir_trees_equal(new_dir1, new_dir2):
            return False
    return True
```

Moreover, `common_files` should return what files are common between the two directories, once copied, and thus, should return all files at the level. This should certainly be > 0 in a successfully copy, so we can't put that in the 'return False' case.

Reply:
Ah I misunderstood this function's purpose, I thought it was to validate that the two directories are the same, not that it was determining whether they were the same or not. And I see the recursion that goes into subdirectories. Thanks for the clarification!